pkg/agent/plugin/svidstore/gcpsecretmanager/gcloud.go

```diff
@@ -0,0 +1,286 @@
+			// Create a policy without conditions and a single binding
+			resp, err := p.client.SetIamPolicy(ctx, &iam.SetIamPolicyRequest{
+				Resource: opt.secretName(),
+				Policy: &iam.Policy{
+					Bindings: []*iam.Binding{
+						{
+							Role:    opt.roleName,
+							Members: []string{opt.serviceAccount},
+						},
+					},
+				},
+			})
+			if err != nil {
+				return nil, status.Errorf(codes.Internal, "failed to set IAM policy to secret: %v", err)
+			}
+			p.log.With("version", resp.Version).With("etag", resp.Etag).With("secret_name", secret.Name).Debug("Secret IAM Policy updated")
+		}
+	}
+
+	secretData, err := svidstore.SecretFromProto(req)
+	if err != nil {
+		return nil, status.Errorf(codes.InvalidArgument, "failed to parse request: %v", err)
+	}
+
+	secretBinary, err := json.Marshal(secretData)
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to marshal payload: %v", err)
+	}
+
+	resp, err := p.client.AddSecretVersion(ctx, &secretmanagerpb.AddSecretVersionRequest{
+		Parent: secret.Name,
+		Payload: &secretmanagerpb.SecretPayload{
+			Data: secretBinary,
+		},
+	})
+	if err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to add secret version: %v", err)
+	}
+
+	p.log.With("state", resp.State).With("name", resp.Name).Debug("Secret payload updated")
+
+	return &svidstorev1.PutX509SVIDResponse{}, nil
+}
+
+// DeleteX509SVID deletes a Secret in the configured Google Cloud Secrets manager
+func (p *SecretManagerPlugin) DeleteX509SVID(ctx context.Context, req *svidstorev1.DeleteX509SVIDRequest) (*svidstorev1.DeleteX509SVIDResponse, error) {
+	opt, err := optionsFromSecretData(req.Metadata)
+	if err != nil {
+		return nil, err
+	}
+
+	secret, err := getSecret(ctx, p.client, opt.secretName(), p.tdHash)
+	if err != nil {
+		return nil, err
+	}
+
+	if secret == nil {
+		p.log.With("secret_name", opt.secretName()).Debug("Secret to delete not found")
+		return &svidstorev1.DeleteX509SVIDResponse{}, nil
+	}
+
+	if err := p.client.DeleteSecret(ctx, &secretmanagerpb.DeleteSecretRequest{
+		Name: secret.Name,
+		Etag: secret.Etag,
+	}); err != nil {
+		return nil, status.Errorf(codes.Internal, "failed to delete secret: %v", err)
+	}
+
+	p.log.With("secret_name", opt.secretName()).Debug("Secret deleted")
+	return &svidstorev1.DeleteX509SVIDResponse{}, nil
+}
+
+// getSecret gets secret from Google Cloud and validates if it has `spire-svid` label, nil if not found
+func getSecret(ctx context.Context, client secretManagerClient, secretName string, td string) (*secretmanagerpb.Secret, error) {
+	secret, err := client.GetSecret(ctx, &secretmanagerpb.GetSecretRequest{
+		Name: secretName,
+	})
+	switch status.Code(err) {
+	case codes.OK:
+		// Verify that secret contains "spire-svid" label and it is enabled
+		if ok := validateLabels(secret.Labels, td); !ok {
+			return nil, status.Errorf(codes.InvalidArgument, "secret does not contain the 'spire-svid' label for trust domain %q", td)
```

Comment:
This error should be updated. Maybe the error should be more opaque in terms of the logic behind this validation and should only mention that the secret is not managed by this SPIRE deployment, and leave the details to the logging.
We may log the label that was expected and the label that was found. I think that would be useful for operators troubleshooting problems.

Reply:
mmm not sure about logs here, we are not providing another logs for errors from anothers plugins we rely on errors for that